numpy/core/numeric.py

```diff
@@ -401,11 +402,17 @@ def asarray(a, dtype=None, order=None):
         Input data, in any form that can be converted to an array.  This
         includes lists, lists of tuples, tuples, tuples of tuples, tuples
         of lists and ndarrays.
-    dtype : data-type, optional
+    dtype : {None, data-type}, optional
         By default, the data-type is inferred from the input data.
-    order : {'C', 'F'}, optional
-        Whether to use row-major ('C') or column-major ('F' for FORTRAN)
-        memory representation.  Defaults to 'C'.
+    order : {'C', 'F', 'A'}, optional
+        Whether to use row-major ('C'), column-major ('F'), or any ('A')
+        memory representation.  Defaults to 'A'.
+    precision: {None, dtype, ndarray, scalar}, optional
```

Comment:
Thinking out loud:

Maybe some way to specify a type which the final type has to be under?
parent_dtype=np.inexact or something? (Well, I guess inexact may not be a
dtype and is rather an abstract scalar type, because for some reason those
are different things. But still, it seems like it'd make sense to take
advantage the abstract portions of the type hierarchy here.)

In fact, while I know I just argued against overloading kwargs :-), but it
would actually make logical sense for dtype=np.inexact to do something like
this. The intuition is that float64 IS-A inexact, therefore if I ask for an
inexact and you give me a float64 then you have done what I asked.

This would also meet Josef's request on the mailing list for a way to say
"numeric, not object, please".

An important issue is that in the approach I'm sketching out in this
comment, we would use the explicit hierarchy/ordering defined by the type
inheritance tree, while result_type uses the implicit hierarchy/ordering
defined by can_cast (I.e. subset relations between representable values).
So while this is kind of a max_dtype suggestion, it isn't actually
symmetric to Chuck's precision/min_dtype suggestion.

I don't really like relying on the explicit type hierarchy -- that's
usually a code smell in Python, and looking ahead to a future where we have
lots more dtypes, can_cast is easier to extend than the type hierarchy.
Maybe a better approach would actually be a max_dtype that's exactly
symmetric, I.e. you'd say max_dtype=complex128 to mean that whatever dtype
we end up with, it had better be one that safely can_cast to complex128.

Well, except that that would not actually solve the original problem, oops
:-).

What do we actually want? This does not seem to be a case where hierarchies
are a neat fit to the desired semantics. Maybe we should treat inexactness
as a kind of flag attached to a dtype, or something like that, and make it
possible to require certain flags? Or a closed-under-true-division flag?

Do we have a list anywhere of the actual cases where we're currently doing
a + 0.0 or whatever that we could look at to see *exactly* what property
the subsequent code is relying on?
On Mar 6, 2015 12:00 AM, "seberg" <notifications@github.com> wrote:

> In numpy/core/numeric.py
> <https://github.com/numpy/numpy/pull/5634#discussion_r25930165>:
>
> >          By default, the data-type is inferred from the input data.
> > -    order : {'C', 'F'}, optional
> > -        Whether to use row-major ('C') or column-major ('F' for FORTRAN)
> > -        memory representation.  Defaults to 'C'.
> > +    order : {'C', 'F', 'A'}, optional
> > +        Whether to use row-major ('C'), column-major ('F'), or any ('A')
> > +        memory representation.  Defaults to 'A'.
> > +    precision: {None, dtype, ndarray, scalar}, optional
>
> I agree, what the scalar does is difficult to understand, but the
> inexact=True logic seems actually the most common one? And it is not
> quite the same as *any* dtype. I once thought about a prepare_arrays
> (something like asarray but which takes *arrays and would do things like
> this, plus give the common result type/cast them all) and I did not even
> think about min_dtype, but only about inexact.
> Was thinking to do that in C, but could sketch up a python implementation
> of something like this to give an idea if that seems like it might be an
> option.
>
> You do not need to mention scalars ;), since they are actually cast to 0-d
> arrays in result_type, but that does *not* fix the issue that the usage,
> which to me currently seems the most useful, is completely hidden away in
> scalar which makes no sense without more explanation. (though it always
> needs explanation). I don't have a solution, but hopefully these are not
> just unhelpful random thoughts :).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5634/files#r25930165>.
>